0/TransactionBuilder.js

```diff
@@ -150,6 +199,27 @@ class TransactionBuilder {
         return this;
     }
 
+    /**
+     * Add a public key and signature to the transaction. This is used only in the case of unsigned transactions (usefull for hardware wallets).
+     * Public Key's /signatures need to be added in the same order as their corresponding inputs.
```

Comment:
This method also doesn't check whether the key `publicKey` is a valid public key

Reply:
Let me check in my latest changes, and lets revisit this after you look those over.

The signature must have the public key, so if we were to reuse the input, we'd have to do something like `input(fa/fs, amount, publickey, signature).`  I did switch over to the method suggested by Paul by requiring either the FA or Fs only.  The pkSignature will search for the appropriate address in the list, so I don't think it is possible for the signatures to get out of order, assuming the constructor is a copy of the original transaction to be signed.  You would still have to have two passes on the builder, the first pass to build the contents, and the second to only add the signature for those contents.

One approach i tinkered with to simply things was provide a callback function to the inputs, so it would look something like 

`input(fa/fs, amount, signaturecallback)`

Everything was nice and clean, but it eventually presented me with an insurmountable problem.  That is, the Ledger device calls are all asynchronous, so my call back needed to be async as well.  Unwrapping promises in the constructor seemed next to impossible given the object freezes at the end of the constructor.  It seemed my changes would end up being way too intrusive,  so in the interest of time I punted with the callback approach. I'm sure someone with superior js skills can help get that approach to work.

Reply:
I can't see a clean way to do this.  Please take a look at the updated code.